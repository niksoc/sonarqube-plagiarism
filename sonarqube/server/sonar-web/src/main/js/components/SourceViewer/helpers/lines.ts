/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { intersection } from 'lodash';

export function optimizeHighlightedSymbols(
  symbolsForLine: string[] = [],
  highlightedSymbols: string[] = []
): string[] | undefined {
  const symbols = intersection(symbolsForLine, highlightedSymbols);

  return symbols.length ? symbols : undefined;
}

export function optimizeLocationMessage(
  highlightedLocationMessage: { index: number; text: string | undefined } | undefined,
  optimizedSecondaryIssueLocations: T.LinearIssueLocation[]
) {
  return highlightedLocationMessage != null &&
    optimizedSecondaryIssueLocations.some(
      location => location.index === highlightedLocationMessage.index
    )
    ? highlightedLocationMessage
    : undefined;
}

export function optimizeSelectedIssue(selectedIssue: string | undefined, issuesForLine: T.Issue[]) {
  return selectedIssue !== undefined && issuesForLine.find(issue => issue.key === selectedIssue)
    ? selectedIssue
    : undefined;
}
