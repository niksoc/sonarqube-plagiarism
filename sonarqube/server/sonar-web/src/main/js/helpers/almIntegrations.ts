/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { isLoggedIn } from './users';

export function getAlmMembersUrl(key: string, url: string): string {
  if (!url.endsWith('/')) {
    url += '/';
  }
  if (isGithub(key)) {
    return url.replace('github.com/', 'github.com/orgs/') + 'people';
  }
  return url + 'profile/members';
}

export function getUserAlmKey(user: T.CurrentUser) {
  return isLoggedIn(user) && user.externalProvider
    ? sanitizeAlmId(user.externalProvider)
    : undefined;
}

export function hasAdvancedALMIntegration(user: T.CurrentUser) {
  return (
    isLoggedIn(user) && (isBitbucket(user.externalProvider) || isGithub(user.externalProvider))
  );
}

export function isBitbucket(almKey?: string): boolean {
  return almKey ? almKey.startsWith('bitbucket') : false;
}

export function isGithub(almKey?: string): boolean {
  return almKey === 'github';
}

export function isVSTS(almKey?: string): boolean {
  return almKey === 'microsoft';
}

export function sanitizeAlmId(almKey: string) {
  if (isBitbucket(almKey)) {
    return 'bitbucket';
  }
  return almKey;
}
