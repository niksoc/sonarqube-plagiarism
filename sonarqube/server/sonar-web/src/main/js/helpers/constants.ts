/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { colors } from '../app/theme';

export const SEVERITIES = ['BLOCKER', 'CRITICAL', 'MAJOR', 'MINOR', 'INFO'];
export const STATUSES = ['OPEN', 'REOPENED', 'CONFIRMED', 'RESOLVED', 'CLOSED'];
export const ISSUE_TYPES: T.IssueType[] = [
  'BUG',
  'VULNERABILITY',
  'CODE_SMELL',
  'SECURITY_HOTSPOT'
];
export const RULE_TYPES: T.RuleType[] = ['BUG', 'VULNERABILITY', 'CODE_SMELL', 'SECURITY_HOTSPOT'];
export const RULE_STATUSES = ['READY', 'BETA', 'DEPRECATED'];

export const CHART_COLORS_RANGE_PERCENT = [
  colors.green,
  colors.lightGreen,
  colors.yellow,
  colors.orange,
  colors.red
];

export const CHART_REVERSED_COLORS_RANGE_PERCENT = [
  colors.red,
  colors.orange,
  colors.yellow,
  colors.lightGreen,
  colors.green
];

export const RATING_COLORS = [
  colors.green,
  colors.lightGreen,
  colors.yellow,
  colors.orange,
  colors.red
];
