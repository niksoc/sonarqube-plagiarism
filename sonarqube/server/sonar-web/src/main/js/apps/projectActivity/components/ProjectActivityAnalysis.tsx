/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import * as React from 'react';
import ActionsDropdown, {
  ActionsDropdownDivider,
  ActionsDropdownItem
} from 'sonar-ui-common/components/controls/ActionsDropdown';
import ClickEventBoundary from 'sonar-ui-common/components/controls/ClickEventBoundary';
import HelpTooltip from 'sonar-ui-common/components/controls/HelpTooltip';
import { PopupPlacement } from 'sonar-ui-common/components/ui/popups';
import { parseDate } from 'sonar-ui-common/helpers/dates';
import { translate, translateWithParameters } from 'sonar-ui-common/helpers/l10n';
import { scrollToElement } from 'sonar-ui-common/helpers/scrolling';
import TimeFormatter from '../../../components/intl/TimeFormatter';
import Events from './Events';
import AddEventForm from './forms/AddEventForm';
import RemoveAnalysisForm from './forms/RemoveAnalysisForm';

export interface ProjectActivityAnalysisProps {
  addCustomEvent: (analysis: string, name: string, category?: string) => Promise<void>;
  addVersion: (analysis: string, version: string) => Promise<void>;
  analysis: T.ParsedAnalysis;
  canAdmin?: boolean;
  canDeleteAnalyses?: boolean;
  canCreateVersion: boolean;
  changeEvent: (event: string, name: string) => Promise<void>;
  deleteAnalysis: (analysis: string) => Promise<void>;
  deleteEvent: (analysis: string, event: string) => Promise<void>;
  isBaseline: boolean;
  isFirst: boolean;
  parentScrollContainer?: HTMLElement | null;
  selected: boolean;
  updateSelectedDate: (date: Date) => void;
}

export function ProjectActivityAnalysis(props: ProjectActivityAnalysisProps) {
  let node: HTMLLIElement | null = null;

  const {
    analysis,
    isBaseline,
    isFirst,
    canAdmin,
    canCreateVersion,
    parentScrollContainer,
    selected
  } = props;

  React.useEffect(() => {
    if (node && parentScrollContainer && selected) {
      const { height } = node.getBoundingClientRect();
      scrollToElement(node, {
        bottomOffset: height + 20,
        topOffset: 60,
        parent: parentScrollContainer,
        smooth: false
      });
    }
  });

  const [addEventForm, setAddEventForm] = React.useState(false);
  const [addVersionForm, setAddVersionForm] = React.useState(false);
  const [removeAnalysisForm, setRemoveAnalysisForm] = React.useState(false);

  const parsedDate = parseDate(analysis.date);
  const hasVersion = analysis.events.find(event => event.category === 'VERSION') != null;

  const canAddVersion = canAdmin && !hasVersion && canCreateVersion;
  const canAddEvent = canAdmin;
  const canDeleteAnalyses =
    props.canDeleteAnalyses && !isFirst && !analysis.manualNewCodePeriodBaseline;

  return (
    <li
      className={classNames('project-activity-analysis bordered-top bordered-bottom', {
        selected
      })}
      onClick={() => props.updateSelectedDate(analysis.date)}
      ref={ref => (node = ref)}>
      <div className="display-flex-center display-flex-space-between">
        <div className="project-activity-time">
          <TimeFormatter date={parsedDate} long={false}>
            {formattedTime => (
              <time className="text-middle" dateTime={parsedDate.toISOString()}>
                {formattedTime}
              </time>
            )}
          </TimeFormatter>
        </div>

        {analysis.buildString && (
          <div className="flex-shrink small text-muted text-ellipsis">
            {translateWithParameters(
              'project_activity.analysis_build_string_X',
              analysis.buildString
            )}
          </div>
        )}

        {(canAddVersion || canAddEvent || canDeleteAnalyses) && (
          <ClickEventBoundary>
            <div className="project-activity-analysis-actions big-spacer-left">
              <ActionsDropdown
                overlayPlacement={PopupPlacement.BottomRight}
                small={true}
                toggleClassName="js-analysis-actions">
                {canAddVersion && (
                  <ActionsDropdownItem
                    className="js-add-version"
                    onClick={() => setAddVersionForm(true)}>
                    {translate('project_activity.add_version')}
                  </ActionsDropdownItem>
                )}
                {canAddEvent && (
                  <ActionsDropdownItem
                    className="js-add-event"
                    onClick={() => setAddEventForm(true)}>
                    {translate('project_activity.add_custom_event')}
                  </ActionsDropdownItem>
                )}
                {(canAddVersion || canAddEvent) && canDeleteAnalyses && <ActionsDropdownDivider />}
                {canDeleteAnalyses && (
                  <ActionsDropdownItem
                    className="js-delete-analysis"
                    destructive={true}
                    onClick={() => setRemoveAnalysisForm(true)}>
                    {translate('project_activity.delete_analysis')}
                  </ActionsDropdownItem>
                )}
              </ActionsDropdown>

              {addVersionForm && (
                <AddEventForm
                  addEvent={props.addVersion}
                  addEventButtonText="project_activity.add_version"
                  analysis={analysis}
                  onClose={() => setAddVersionForm(false)}
                />
              )}

              {addEventForm && (
                <AddEventForm
                  addEvent={props.addCustomEvent}
                  addEventButtonText="project_activity.add_custom_event"
                  analysis={analysis}
                  onClose={() => setAddEventForm(false)}
                />
              )}

              {removeAnalysisForm && (
                <RemoveAnalysisForm
                  analysis={analysis}
                  deleteAnalysis={props.deleteAnalysis}
                  onClose={() => setRemoveAnalysisForm(false)}
                />
              )}
            </div>
          </ClickEventBoundary>
        )}
      </div>

      {analysis.events.length > 0 && (
        <Events
          analysisKey={analysis.key}
          canAdmin={canAdmin}
          events={analysis.events}
          isFirst={isFirst}
          onChange={props.changeEvent}
          onDelete={props.deleteEvent}
        />
      )}

      {isBaseline && (
        <div className="baseline-marker">
          <div className="wedge" />
          <hr />
          <div className="label display-flex-center">
            {translate('project_activity.new_code_period_start')}
            <HelpTooltip
              className="little-spacer-left"
              overlay={translate('project_activity.new_code_period_start.help')}
              placement="top"
            />
          </div>
        </div>
      )}
    </li>
  );
}

export default React.memo(ProjectActivityAnalysis);
