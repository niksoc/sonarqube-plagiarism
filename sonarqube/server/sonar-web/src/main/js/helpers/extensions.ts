/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { getBaseUrl } from 'sonar-ui-common/helpers/urls';
import { getExtensionFromCache } from './extensionsHandler';

let librariesExposed = false;

export function installScript(url: string, target: 'body' | 'head' = 'body'): Promise<any> {
  return new Promise(resolve => {
    const scriptTag = document.createElement('script');
    scriptTag.src = `${getBaseUrl()}${url}`;
    scriptTag.onload = resolve;
    document.getElementsByTagName(target)[0].appendChild(scriptTag);
  });
}

export async function getExtensionStart(key: string) {
  const fromCache = getExtensionFromCache(key);
  if (fromCache) {
    return Promise.resolve(fromCache);
  }

  if (!librariesExposed) {
    // Async import allows to reduce initial vendor bundle size
    const exposeLibraries = (await import('../app/components/extensions/exposeLibraries')).default;
    exposeLibraries();
    librariesExposed = true;
  }

  await installScript(`/static/${key}.js`);

  const start = getExtensionFromCache(key);
  if (start) {
    return start;
  }
  return Promise.reject();
}
