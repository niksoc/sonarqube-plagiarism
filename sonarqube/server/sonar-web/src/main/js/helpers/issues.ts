/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { flatten, sortBy } from 'lodash';
import { ISSUE_TYPES } from './constants';

interface Comment {
  login: string;
  [x: string]: any;
}

interface Rule {}

interface Component {
  key: string;
  name: string;
}

interface IssueBase {
  severity: string;
  [x: string]: any;
}

export interface RawIssue extends IssueBase {
  assignee?: string;
  author?: string;
  comments?: Array<Comment>;
  component: string;
  flows?: Array<{
    // `componentName` is not available in RawIssue
    locations?: Array<T.Omit<T.FlowLocation, 'componentName'>>;
  }>;
  key: string;
  line?: number;
  project: string;
  rule: string;
  status: string;
  subProject?: string;
  textRange?: T.TextRange;
}

export function sortByType<T extends Pick<T.Issue, 'type'>>(issues: T[]): T[] {
  return sortBy(issues, issue => ISSUE_TYPES.indexOf(issue.type));
}

function injectRelational(
  issue: T.Dict<any>,
  source: any[] | undefined,
  baseField: string,
  lookupField: string
) {
  const newFields: T.Dict<any> = {};
  const baseValue = issue[baseField];
  if (baseValue !== undefined && source !== undefined) {
    const lookupValue = source.find(candidate => candidate[lookupField] === baseValue);
    if (lookupValue != null) {
      Object.keys(lookupValue).forEach(key => {
        const newKey = baseField + key.charAt(0).toUpperCase() + key.slice(1);
        newFields[newKey] = lookupValue[key];
      });
    }
  }
  return newFields;
}

function injectCommentsRelational(issue: RawIssue, users?: T.UserBase[]) {
  if (!issue.comments) {
    return {};
  }
  const comments = issue.comments.map(comment => {
    const commentWithAuthor = { ...comment, author: comment.login, login: undefined };
    return {
      ...commentWithAuthor,
      ...injectRelational(commentWithAuthor, users, 'author', 'login')
    };
  });
  return { comments };
}

function prepareClosed(
  issue: RawIssue,
  secondaryLocations: T.FlowLocation[],
  flows: T.FlowLocation[][]
) {
  return issue.status === 'CLOSED'
    ? { flows: [], line: undefined, textRange: undefined, secondaryLocations: [] }
    : { flows, secondaryLocations };
}

function ensureTextRange(issue: RawIssue): { textRange?: T.TextRange } {
  return issue.line && !issue.textRange
    ? {
        textRange: {
          startLine: issue.line,
          endLine: issue.line,
          startOffset: 0,
          endOffset: 999999
        }
      }
    : {};
}

function reverseLocations(locations: T.FlowLocation[]): T.FlowLocation[] {
  const x = [...locations];
  x.reverse();
  return x;
}

function splitFlows(
  issue: RawIssue,
  components: Component[] = []
): { secondaryLocations: T.FlowLocation[]; flows: T.FlowLocation[][] } {
  const parsedFlows: T.FlowLocation[][] = (issue.flows || [])
    .filter(flow => flow.locations !== undefined)
    .map(flow => flow.locations!.filter(location => location.textRange != null))
    .map(flow =>
      flow.map(location => {
        const component = components.find(component => component.key === location.component);
        return { ...location, componentName: component && component.name };
      })
    );

  const onlySecondaryLocations = parsedFlows.every(flow => flow.length === 1);

  return onlySecondaryLocations
    ? { secondaryLocations: orderLocations(flatten(parsedFlows)), flows: [] }
    : { secondaryLocations: [], flows: parsedFlows.map(reverseLocations) };
}

function orderLocations(locations: T.FlowLocation[]) {
  return sortBy(
    locations,
    location => location.textRange && location.textRange.startLine,
    location => location.textRange && location.textRange.startOffset
  );
}

export function parseIssueFromResponse(
  issue: RawIssue,
  components?: Component[],
  users?: T.UserBase[],
  rules?: Rule[]
): T.Issue {
  const { secondaryLocations, flows } = splitFlows(issue, components);
  return {
    ...issue,
    ...injectRelational(issue, components, 'component', 'key'),
    ...injectRelational(issue, components, 'project', 'key'),
    ...injectRelational(issue, components, 'subProject', 'key'),
    ...injectRelational(issue, rules, 'rule', 'key'),
    ...injectRelational(issue, users, 'assignee', 'login'),
    ...injectCommentsRelational(issue, users),
    ...prepareClosed(issue, secondaryLocations, flows),
    ...ensureTextRange(issue)
  } as T.Issue;
}
