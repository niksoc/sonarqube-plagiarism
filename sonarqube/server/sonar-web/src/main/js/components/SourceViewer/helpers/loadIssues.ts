/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { searchIssues } from '../../../api/issues';
import { getBranchLikeQuery } from '../../../helpers/branch-like';
import { parseIssueFromResponse } from '../../../helpers/issues';
import { BranchLike } from '../../../types/branch-like';

// maximum possible value
const PAGE_SIZE = 500;

function buildQuery(component: string, branchLike: BranchLike | undefined) {
  return {
    additionalFields: '_all',
    resolved: 'false',
    componentKeys: component,
    s: 'FILE_LINE',
    ...getBranchLikeQuery(branchLike)
  };
}

export function loadPage(
  query: T.RawQuery,
  page: number,
  pageSize = PAGE_SIZE
): Promise<T.Issue[]> {
  return searchIssues({
    ...query,
    p: page,
    ps: pageSize
  }).then(r =>
    r.issues.map(issue => parseIssueFromResponse(issue, r.components, r.users, r.rules))
  );
}

export function loadPageAndNext(
  query: T.RawQuery,
  toLine: number,
  page: number,
  pageSize = PAGE_SIZE
): Promise<T.Issue[]> {
  return loadPage(query, page).then(issues => {
    if (issues.length === 0) {
      return [];
    }

    const lastIssue = issues[issues.length - 1];

    if (
      (lastIssue.textRange != null && lastIssue.textRange.endLine > toLine) ||
      issues.length < pageSize
    ) {
      return issues;
    }

    return loadPageAndNext(query, toLine, page + 1, pageSize).then(nextIssues => {
      return [...issues, ...nextIssues];
    });
  });
}

export default function loadIssues(
  component: string,
  _fromLine: number,
  toLine: number,
  branchLike: BranchLike | undefined
): Promise<T.Issue[]> {
  const query = buildQuery(component, branchLike);
  return new Promise(resolve => {
    loadPageAndNext(query, toLine, 1).then(issues => {
      resolve(issues);
    });
  });
}
