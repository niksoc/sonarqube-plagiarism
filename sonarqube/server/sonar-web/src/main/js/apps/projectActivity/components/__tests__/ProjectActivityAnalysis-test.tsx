/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
/* eslint-disable sonarjs/no-duplicate-string */
import { mount, shallow } from 'enzyme';
import * as React from 'react';
import { IntlProvider } from 'react-intl';
import { scrollToElement } from 'sonar-ui-common/helpers/scrolling';
import { click } from 'sonar-ui-common/helpers/testUtils';
import TimeFormatter from '../../../../components/intl/TimeFormatter';
import { mockAnalysisEvent, mockParsedAnalysis } from '../../../../helpers/testMocks';
import AddEventForm from '../forms/AddEventForm';
import RemoveAnalysisForm from '../forms/RemoveAnalysisForm';
import { ProjectActivityAnalysis, ProjectActivityAnalysisProps } from '../ProjectActivityAnalysis';

jest.mock('sonar-ui-common/helpers/dates', () => ({
  parseDate: () => ({
    valueOf: () => 1546333200000,
    toISOString: () => '2019-01-01T09:00:00.000Z'
  })
}));

jest.mock('sonar-ui-common/helpers/scrolling', () => ({
  scrollToElement: jest.fn()
}));

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot('default');
  expect(
    shallowRender({ analysis: mockParsedAnalysis({ events: [mockAnalysisEvent()] }) })
  ).toMatchSnapshot('with events');
  expect(
    shallowRender({ analysis: mockParsedAnalysis({ buildString: '1.0.234' }) })
  ).toMatchSnapshot('with build string');
  expect(shallowRender({ isBaseline: true })).toMatchSnapshot('with baseline marker');
  expect(
    shallowRender({
      canAdmin: true,
      canCreateVersion: true,
      canDeleteAnalyses: true
    })
  ).toMatchSnapshot('with admin options');

  const timeFormatter = shallowRender()
    .find(TimeFormatter)
    .prop('children');
  if (!timeFormatter) {
    fail('TimeFormatter instance not found');
  } else {
    expect(timeFormatter('formatted_time')).toMatchSnapshot('formatted time');
  }
});

it('should show the correct admin options', () => {
  const wrapper = shallowRender({
    canAdmin: true,
    canCreateVersion: true,
    canDeleteAnalyses: true
  });

  expect(wrapper.find('.js-add-version').exists()).toBe(true);
  click(wrapper.find('.js-add-version'));
  const addVersionForm = wrapper.find(AddEventForm);
  expect(addVersionForm.exists()).toBe(true);
  addVersionForm.prop('onClose')();
  expect(wrapper.find(AddEventForm).exists()).toBe(false);

  expect(wrapper.find('.js-add-event').exists()).toBe(true);
  click(wrapper.find('.js-add-event'));
  const addEventForm = wrapper.find(AddEventForm);
  expect(addEventForm.exists()).toBe(true);
  addEventForm.prop('onClose')();
  expect(wrapper.find(AddEventForm).exists()).toBe(false);

  expect(wrapper.find('.js-delete-analysis').exists()).toBe(true);
  click(wrapper.find('.js-delete-analysis'));
  const removeAnalysisForm = wrapper.find(RemoveAnalysisForm);
  expect(removeAnalysisForm.exists()).toBe(true);
  removeAnalysisForm.prop('onClose')();
  expect(wrapper.find(RemoveAnalysisForm).exists()).toBe(false);
});

it('should not allow the first item to be deleted', () => {
  expect(
    shallowRender({
      canAdmin: true,
      canCreateVersion: true,
      canDeleteAnalyses: true,
      isFirst: true
    })
      .find('.js-delete-analysis')
      .exists()
  ).toBe(false);
});

it('should be clickable', () => {
  const date = new Date('2018-03-01T09:37:01+0100');
  const updateSelectedDate = jest.fn();
  const wrapper = shallowRender({ analysis: mockParsedAnalysis({ date }), updateSelectedDate });
  click(wrapper);
  expect(updateSelectedDate).toBeCalledWith(date);
});

it('should trigger a scroll to itself if selected', () => {
  mountRender({ parentScrollContainer: document.createElement('ul'), selected: true });
  expect(scrollToElement).toBeCalled();
});

function shallowRender(props: Partial<ProjectActivityAnalysisProps> = {}) {
  return shallow(createComponent(props));
}

function mountRender(props: Partial<ProjectActivityAnalysisProps> = {}) {
  return mount(<IntlProvider locale="en">{createComponent(props)}</IntlProvider>);
}

function createComponent(props: Partial<ProjectActivityAnalysisProps> = {}) {
  return (
    <ProjectActivityAnalysis
      addCustomEvent={jest.fn()}
      addVersion={jest.fn()}
      analysis={mockParsedAnalysis()}
      canCreateVersion={false}
      changeEvent={jest.fn()}
      deleteAnalysis={jest.fn()}
      deleteEvent={jest.fn()}
      isBaseline={false}
      isFirst={false}
      selected={false}
      updateSelectedDate={jest.fn()}
      {...props}
    />
  );
}
