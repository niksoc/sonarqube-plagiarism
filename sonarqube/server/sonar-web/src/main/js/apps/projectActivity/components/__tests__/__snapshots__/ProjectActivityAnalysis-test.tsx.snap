// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly: default 1`] = `
<li
  className="project-activity-analysis bordered-top bordered-bottom"
  onClick={[Function]}
>
  <div
    className="display-flex-center display-flex-space-between"
  >
    <div
      className="project-activity-time"
    >
      <TimeFormatter
        date={
          Object {
            "toISOString": [Function],
            "valueOf": [Function],
          }
        }
        long={false}
      >
        <Component />
      </TimeFormatter>
    </div>
  </div>
</li>
`;

exports[`should render correctly: formatted time 1`] = `
<time
  className="text-middle"
  dateTime="2019-01-01T09:00:00.000Z"
>
  formatted_time
</time>
`;

exports[`should render correctly: with admin options 1`] = `
<li
  className="project-activity-analysis bordered-top bordered-bottom"
  onClick={[Function]}
>
  <div
    className="display-flex-center display-flex-space-between"
  >
    <div
      className="project-activity-time"
    >
      <TimeFormatter
        date={
          Object {
            "toISOString": [Function],
            "valueOf": [Function],
          }
        }
        long={false}
      >
        <Component />
      </TimeFormatter>
    </div>
    <ClickEventBoundary>
      <div
        className="project-activity-analysis-actions big-spacer-left"
      >
        <ActionsDropdown
          overlayPlacement="bottom-right"
          small={true}
          toggleClassName="js-analysis-actions"
        >
          <ActionsDropdownItem
            className="js-add-version"
            onClick={[Function]}
          >
            project_activity.add_version
          </ActionsDropdownItem>
          <ActionsDropdownItem
            className="js-add-event"
            onClick={[Function]}
          >
            project_activity.add_custom_event
          </ActionsDropdownItem>
          <ActionsDropdownDivider />
          <ActionsDropdownItem
            className="js-delete-analysis"
            destructive={true}
            onClick={[Function]}
          >
            project_activity.delete_analysis
          </ActionsDropdownItem>
        </ActionsDropdown>
      </div>
    </ClickEventBoundary>
  </div>
</li>
`;

exports[`should render correctly: with baseline marker 1`] = `
<li
  className="project-activity-analysis bordered-top bordered-bottom"
  onClick={[Function]}
>
  <div
    className="display-flex-center display-flex-space-between"
  >
    <div
      className="project-activity-time"
    >
      <TimeFormatter
        date={
          Object {
            "toISOString": [Function],
            "valueOf": [Function],
          }
        }
        long={false}
      >
        <Component />
      </TimeFormatter>
    </div>
  </div>
  <div
    className="baseline-marker"
  >
    <div
      className="wedge"
    />
    <hr />
    <div
      className="label display-flex-center"
    >
      project_activity.new_code_period_start
      <HelpTooltip
        className="little-spacer-left"
        overlay="project_activity.new_code_period_start.help"
        placement="top"
      />
    </div>
  </div>
</li>
`;

exports[`should render correctly: with build string 1`] = `
<li
  className="project-activity-analysis bordered-top bordered-bottom"
  onClick={[Function]}
>
  <div
    className="display-flex-center display-flex-space-between"
  >
    <div
      className="project-activity-time"
    >
      <TimeFormatter
        date={
          Object {
            "toISOString": [Function],
            "valueOf": [Function],
          }
        }
        long={false}
      >
        <Component />
      </TimeFormatter>
    </div>
    <div
      className="flex-shrink small text-muted text-ellipsis"
    >
      project_activity.analysis_build_string_X.1.0.234
    </div>
  </div>
</li>
`;

exports[`should render correctly: with events 1`] = `
<li
  className="project-activity-analysis bordered-top bordered-bottom"
  onClick={[Function]}
>
  <div
    className="display-flex-center display-flex-space-between"
  >
    <div
      className="project-activity-time"
    >
      <TimeFormatter
        date={
          Object {
            "toISOString": [Function],
            "valueOf": [Function],
          }
        }
        long={false}
      >
        <Component />
      </TimeFormatter>
    </div>
  </div>
  <Memo(Events)
    analysisKey="foo"
    events={
      Array [
        Object {
          "category": "QUALITY_GATE",
          "description": "Lorem ipsum dolor sit amet",
          "key": "E11",
          "name": "Lorem ipsum",
          "qualityGate": Object {
            "failing": Array [
              Object {
                "branch": "master",
                "key": "foo",
                "name": "Foo",
              },
              Object {
                "branch": "feature/bar",
                "key": "bar",
                "name": "Bar",
              },
            ],
            "status": "ERROR",
            "stillFailing": true,
          },
        },
      ]
    }
    isFirst={false}
    onChange={[MockFunction]}
    onDelete={[MockFunction]}
  />
</li>
`;
